src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private LowLevelConcurrentStack<SafeHeapHandle> handleCache = new LowLevelConcurrentStack<SafeHeapHandle>();
```

Comment:
> How are you calculating 1MB?

Sorry 1K.

> Can you please share the benchmark you used?

The loop you provided, both as is and swapping with Parallel.For with varying max conc. Against your relative and full path as well as a shorter full path. Release build, x64 with a copy of the original desktop code.

> We currently use scheme A. 

We use untested A. Again, I completely rewrote this. B is more appropriate. We're establishing a new baseline.

> I didn't say anything about zeroing.

I'm pointing it out because it is part of what I'm leveraging.

> I'm not saying you're making arbitrary changes. I'm saying your'e making a whole bag of changes all together, and in doing so it's not clear that each individual change is a net win. I'm saying that maintaining a global, never-shrinking pool of buffers can be quite wasteful. I'm saying that we've significantly increased the amount of unsafe code being used, and I'm concerned we're introduced subtle bugs that will be difficult to catch. I'm saying that these changes introduce additional allocations that are difficult to quantify as part of microbenchmarks but that can have an effect under real-world load.

If I hadn't *completely* rewritten the code I would understand the desire to make slower iterations. This is a completely new baseline. I've tested each and every variant of the individual pieces and found a win for every one. In many cases at the win is at least twice as fast. How can I do better than that? We have no idea how a brand new algorithm will impact real-world scenarios. Do the best we can to start using the legacy logic to inform, but wait for actual measurements to address rather than guessing, however well informed.

I do know from extensive experience with Path.GetFullPath as a client that the perf of normalizing a normalized path kills. It was the single biggest hit for the project system in Blend. I know that the new code is at least twice as fast here and I know it scales far better to longer paths.

I also know this design aligns with other work around buffers and could inform / adapt to / consume it.